lib/cache/OperationWriter.js

```diff
@@ -0,0 +1,255 @@
+      return;
+    }
+    // cancel any scheduled flush
+    if(timeoutId !== null) {
+      clearTimeout(timeoutId);
+      this.timeoutId = null;
+    }
+    this.running = false;
+    this.stopped = null;
+    this._resolveStopped();
+  }
+
+  /**
+   * Adds an operation to be written. If this OperationWriter is running and the
+   * operation causes the queue to reach the configured `highWaterMark`, then
+   * it will automatically flush operations to the underlying database via a
+   * bulk insert.
+   *
+   * @param {object} options - Options to use.
+   * @param {object} options.operation - The operation to write.
+   * @param {object} options.meta - The metadata to write.
+   */
+  async add({operation, meta, forceFlush = false} = {}) {
+    const {queue, running, highWaterMark, MAX_OPERATION_QUEUE_LENGTH} = this;
+
+    if(this.databaseOperationQueueLength >= MAX_OPERATION_QUEUE_LENGTH) {
+      // we intentionally do not await the promise, the length can be updated
+      // after the error is returned
+      this._cacheDatabaseQueueLength().catch(e => {
+        logger.debug(DATABASE_QUEUE_LENGTH_ERROR, {error: e});
+      });
+      throw new BedrockError(
+        'The node is not accepting operations. Try again later.',
+        'OperationError', {
+          httpStatusCode: 503,
+          public: true,
+        });
+    }
+
+    if(this._isDuplicateOperation(meta)) {
+      throw new BedrockError(
+        'The operation already exists.',
+        'DuplicateError', {
+          duplicateLocation: 'cache',
+          httpStatusCode: 409,
+          ledgerNodeId: this.ledgerNodeId,
+          operationRecord: {operation, meta},
+          public: true
+        });
+    }
+
+    // queue the operation
+    queue.push({operation, meta});
+
+    // do not flush if not writer is not running
+    if(!running) {
+      return;
+    }
+
+    // force flush if required
+    if(forceFlush) {
+      await this.flush();
+    }
+
+    // if the queue length matches the highWaterMark, schedule an immediate
+    // flush; only schedule this when they match do prevent repeatedly
+    // scheduling flushes or preventing flushes under heavy load
+    if(queue.length === highWaterMark) {
+      this._scheduleFlush({timeout: 0});
+      return;
+    }
+
+    // schedule a flush if one has not already been scheduled
+    const {timeout, timeoutId} = this;
+    if(timeoutId === null) {
+      this._scheduleFlush({timeout});
+    }
+  }
+
+  /**
+   * Forcibly flushes the operation queue, bulk inserting any pending
+   * operations.
+   *
+   * @returns {Promise} A promise that resolves once the operation completes.
+   */
+  async flush() {
+    // clear any scheduled flush
+    if(this.timeoutId !== null) {
+      clearTimeout(this.timeoutId);
+      this.timeoutId = null;
+    }
+
+    // save current queue and replace it with a new one
+    const queue = this.queue;
+    this.queue = [];
+
+    if(queue.length === 0) {
+      return;
+    }
```

Comment:
Should checking the queue length happen first so as to avoid instantiating a new array unnecessarily?

```suggestion
    if(queue.length === 0) {
      return;
    }

    // save current queue and replace it with a new one
    const queue = this.queue;
    this.queue = [];
```